konstraints-core/src/test/kotlin/tools/aqua/konstraints/ParseContextTests.kt

```diff
@@ -0,0 +1,325 @@
+
+    assertEquals(sort, function?.sort)
+  }
+
+  @ParameterizedTest
+  @MethodSource("getFunctionNameAndArgs")
+  @Order(4)
+  fun `Check that Function accepts parameters`(
+      name: String,
+      args: List<Expression<*>>,
+      sort: Sort
+  ) {
+    val function = context.getFunction(name, args)
+
+    requireNotNull(function)
+    assert(function.acceptsExpressions(args, emptyList()))
+  }
+
+  private fun getFunctionNameAndArgs(): Stream<Arguments> {
+    return Stream.of(arguments("O", listOf(boolExpression, boolExpression), BoolSort))
+  }
+
+  @ParameterizedTest
+  @MethodSource("getFunctionNameAndIncorrectArgs")
+  @Order(5)
+  fun `Check that Function rejects incorrect parameters`(
+      function: FunctionDecl<*>,
+      args: List<Expression<*>>
+  ) {
+    assertFalse(function.acceptsExpressions(args, emptyList()))
+  }
+
+  @ParameterizedTest
+  @MethodSource("getFunctionNameAndIncorrectArgs")
+  @Order(6)
+  fun `Check that Expression generation fails for incorrect parameters`(
+      function: FunctionDecl<*>,
+      args: List<Expression<*>>
+  ) {
+    assertThrows<Exception> { function.buildExpression(args, emptyList()) }
+  }
+
+  private fun getFunctionNameAndIncorrectArgs(): Stream<Arguments> {
+    return Stream.of(
+        arguments(AndDecl, listOf(boolExpression, boolExpression, bv16Expression)),
+        arguments(OrDecl, listOf(boolExpression, boolExpression, bv16Expression)),
+        arguments(XOrDecl, listOf(boolExpression, boolExpression, bv16Expression)),
+        arguments(BVUltDecl, listOf(bv16Expression)),
+        arguments(BVUltDecl, listOf(boolExpression, boolExpression)),
+        arguments(BVUltDecl, listOf(bv16Expression, boolExpression)),
+        arguments(BVUltDecl, listOf(bv16Expression, bv32Expression)),
+        arguments(BVUltDecl, listOf(bv32Expression, bv16Expression)),
+        arguments(overloadedBV, listOf(bv16Expression, bv16Expression)),
+        arguments(EqualsDecl, listOf(boolExpression, bv16Expression)),
+        arguments(EqualsDecl, listOf(boolExpression, boolExpression, bv16Expression)),
+        arguments(EqualsDecl, listOf(boolExpression, bv16Expression, boolExpression)))
+  }
+
+  /*
+   * This test must be executed after all of the above tests, as the function symbol "and" will become ambiguous
+   * making it illegal to use outside "match" and "as"
+   */
+  @ParameterizedTest
+  @MethodSource("getFunctionDeclarations")
+  @Order(7)
+  fun `test that legal functions can be registered`(func: FunctionDecl<*>) {
+    context.registerFunction(func)
+  }
+
+  private fun getFunctionDeclarations(): Stream<Arguments> {
+    return Stream.of(
+        /* No conflict */
+        arguments(
+            FunctionDecl(
+                "foo".symbol(),
+                emptySet(),
+                listOf(BoolSort, BoolSort),
+                emptySet(),
+                emptySet(),
+                BoolSort,
+                Associativity.NONE,
+                null)))
```

Comment:
the parameter list is far too long; nobody can understand what happens here from the caller side, I fear